cirq/ops/linear_combinations.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Dict, Mapping, Optional, Tuple
+
+import numpy as np
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+
+
+class LinearCombinationOfGates(value.LinearDict[raw_types.Gate]):
+    """Represents linear operator defined by a linear combination of gates.
+
+    Suppose G1, G2, ..., Gn are gates and b1, b2, ..., bn are complex
+    numbers. Then
+
+        LinearCombinationOfGates({G1: b1, G2: b2, ..., Gn: bn})
+
+    represents the linear operator
+
+        A = b1 G1 + b2 G2 + ... + bn * Gn
+
+    Note that A may not be unitary or even normal.
+
+    Rather than creating LinearCombinationOfGates instance explicitly, one may
+    use overloaded arithmetic operators. For example,
+
+        cirq.LinearCombinationOfGates({cirq.X: 2, cirq.Z: -2})
+
+    is equivalent to
+
+        2 * cirq.X - 2 * cirq.Z
+    """
+    def __init__(self, terms: Mapping[raw_types.Gate, value.Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of gates to coefficients in the linear combination
+                being initialized.
+        """
+        super().__init__(terms, validator=self._is_compatible)
+
+    def num_qubits(self) -> Optional[int]:
+        """Returns number of qubits in the domain if known, None if unknown."""
+        if not self:
+            return None
+        any_gate = next(iter(self))
+        return any_gate.num_qubits()
+
+    def _is_compatible(self, gate: raw_types.Gate) -> bool:
+        return (self.num_qubits() is None or
+                self.num_qubits() == gate.num_qubits())
+
+    def matrix(self) -> np.ndarray:
+        num_qubits = self.num_qubits()
+        if num_qubits is None:
+            raise ValueError('Unknown number of qubits')
+        num_dim = 2 ** num_qubits
+        result = np.zeros((num_dim, num_dim), dtype=np.complex128)
+        for gate, coefficient in self.items():
+            result += protocols.unitary(gate) * coefficient
```

Comment:
We've been loosy goosy with dtypes.  I think you would want to make sure that unitary has type that is compatible with summing against a complex128.

Reply:
We have indeed.

I've chosen `np.complex128` assuming that it's safe against all types we use in practice, i.e. that all our unitaries can be promoted to complex128 (and hence can be added into a complex128 array). Tests I did suggest this is the case. Am I missing a type which isn't safe?

An alternative would be to use [`np.result_type()`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.result_type.html#numpy.result_type) to explicitly determine the output type, but AFAICT this will never be a type that cannot be promoted to `np.complex128` anyway.

Reply:
We can't properly ask for a dtype without also adding it as a parameter to `cirq.unitary`. But that would make it harder for people to implement it..

Reply:
I think the strange case is if someone returns complex256 for unitary?

Maybe protocols.unitary should always upconvert to complex128?